Remove link layer reserve concept from network stack

tbursztyka opened the pull request:
This permits to fully separate l2 to l3, finally.

Commit:
net/pkt: Add a pointer to link-layer data in the structure

This is one of the first steps to get rid of ll_reserve attribute and
the related L2 reserve() function.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/ethernet: Let's remove the use for ll reserve

There is no need to reserve any space for each frag, as the l2 will
allocate a frag for the ethernet header, arp will do the same.

This is one step further to removing the concept of ll reserve.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/pkt: Let's ignore ll reserve and use 0 instead

ll reserve is of no use as the l2 is allocating what it needs for
filling in its header.

This is another step forward to removing ll reserve concept.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/ipv4: Remove any use of ll reserve

As this is now unnecessary, let's ignore that and use 0 instead.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/ipv6: Remove any use of ll reserve

As this is now unnecessary, let's ignore that and use 0 instead.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
tests/net: Get rid of the ll reserve

As it is unused now, let's default to 0.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/if: Remove function to get ll reserve

As it is unused now, it can be safely removed.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/l2: Finally get rid of reserve concept from L2

Now that net core does not use ll reserve, it can be nuked from L2.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/pkt: Remove ll_reserve attribute and related functions

It's useless now.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/pkt: Remove parameters to "reserve" some headroom

Such parameter is not used anymore, it was defaulted to 0 previously.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/pkt: Rename net_pkt_ll_clear for better relevancy

Now it does not mangle with any ll reserver space, let's rename it to
net_pkt_lladdr_clear instead.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/6lo: Remove fragmentation function parameter on compression

Fragmentation has nothing to do with 6lo. Up to the bearer to do so.

This change is required for futur serialization of 15.4 fragmentation.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/ieee802154: Serialize packet fragmentation

This will permit to avoid the need for ll_reserve.
It also removes net_buf allocation and deletion, as it works per-frame.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/ieee802154: Finally removing usage of ll_reserve in L2

Moving towards serialized fragmentation, adapting mac command creation
to avoid the need of ll_reserve etc...

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/wifi: Remove usage of ll_reserve

It's not used anymore by the core IP.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

tbursztyka commented:
There are 17 patch on top of PR #10547 
Still some tests that fails, which I am fixing.

rveerama1 commented on subsys/net/l2/ieee802154/ieee802154_fragment.h:
```diff
@@ -18,19 +18,51 @@
- *  @param Pointer to network packet
- *  @param Header difference between original IPv6 header and compressed header
+ *  @param Pointer to valid fragmentatian context
+ *  @param Pointer to varid buffer where to write the fragment
```
Typo s/varid/valid

tbursztyka replied:
and fragmentation instead fragment...meeeh...martian? :]

Commit:
usb/netusb: Let's remove the use for ll reserve

It's useless now.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
samples/net: Let's remove the use of ll reserve

This is now useless.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
tests/ieee80154: Apply L2 changes

Now that fragmentation is serialized and ll_reserve has disappeared from
15.4 L2.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
tests/net: Switch ieee802154 tests to ztest

These were not ran anymore.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

pfalcon commented:
Please rebase.

pfalcon commented on samples/net/throughput_server/prj_frdm_k64f.conf:
```diff
@@ -16,7 +16,7 @@ CONFIG_PRINTK=y
 CONFIG_NET_PKT_RX_COUNT=500
 CONFIG_NET_PKT_TX_COUNT=5
-CONFIG_NET_BUF_RX_COUNT=720
+CONFIG_NET_BUF_RX_COUNT=710
```
Is this change really *required* in this commit? I'd say it only bloats up it, this value could easily stay as is. (Mostly rhetoric note, not worth changing, but IMHO not worth to have been included in the first place.)

tbursztyka replied:
It is required to get that test pass for frdm_k64f. net_pkt has grown a bit, thus this issue.
It's growing even more in next PR #11775, but it's all temporary: in the end it will shrink.

pfalcon replied:
> It is required to get that test pass for frdm_k64f

ack, thanks for explaining

jukkar replied:
Shouldn't we change the CONFIG_NET_PKT_RX_COUNT instead here as that is tied to net_pkt? Probably does not matter much actually.

pfalcon commented on tests/net/dhcpv4/src/main.c:
```diff
@@ -208,14 +208,16 @@ static struct net_buf *pkt_get_data(struct net_pkt *pkt, struct net_if *iface)
+	net_pkt_frag_add(pkt, frag);
+	net_pkt_set_ll(pkt, frag->data);
+
+	hdr = (struct net_eth_hdr *)(net_pkt_ll(pkt));
```
Oh no! No extra parens around function call to be cast! ;-)

pfalcon commented:
Commit  "net/6lo: Remove fragmentation function parameter on compression":

> This change is required for futur serialization of 15.4 fragmentation.

Typo - "futur". Actually, I'm not even sure what "serialization of 15.4 fragmentation" means too.

pfalcon commented on subsys/net/l2/ieee802154/ieee802154_fragment.c:
```diff
@@ -135,166 +115,99 @@ static inline void set_up_frag_hdr(struct net_buf *frag, u16_t size,
- *  fragment at beginning of pkt, move data from next fragments to
- *  previous one, from here on insert fragmentation header and adjust
- *  data on subsequent packets.
+ *  If it's the first fragment being created, fh will not own any offset
```
Is "own" the right word here?

pfalcon reviewed (approved):
It's hard to not notice how this simplifies the code. I'm anxious and eager to see what comes out of all this, so +1 towards merging it sooner rather than later, now it's good timing, to make sure this got properly assimilated and tested before the release.

There were few minor issues noted, I hope they will be fixed before merge.

The biggest I'd like to raise however is in the "Remove link layer reserve concept from network stack" commit. My understanding is that net_pkt_get_reserve_rx() and friends were named such exactly because they were getting a pkt with some ll reserve applied. Now that the reserve is gone, names sounds strange would make sense to not just remove the "reserve" param, but also rename at the same time.

Oh, and as noted by @jukkar in the other review, thanks for splitting the commit well - while the number of them is indeed not trivial, each is sufficiently self-contained to be realistically reviewable. That's noticeable improvement over some previous massive PRs, thanks for reacting to feedback on them.

Commit:
net/ethernet: Remove usage of net_pkt_ll() function

This function is planned to be removed, thus avoiding its usage.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/pkt: Introducing a new helper to get the start of the buffer

This is mostly meant to be used in drivers and/or L2.

net_pkt_ll() is going to be removed and is not semantically right.
Also, net_pkt_ip_data() is not semantically right as it is meant to
access IP data start.

So instead, adding a new function: net_pkt_data() which will get the
start of the buffer.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/ieee802154: Remove usage of net_pkt_ll() function

This function is planned to be removed, thus avoiding its usage.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
tests/net: Remove usage of net_pkt_ll() function

This function is planned to be removed, thus avoiding its usage.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/pkt: Finally removing net_pkt_ll() and the relevant attribute

Removing net_pkt_set_ll() as well of course.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

pfalcon commented:
@tbursztyka:

> but it's a lengthy task

Ack, I agree. Thanks for other hints too.

jukkar commented:
Just fyi, I did not investigate this further yet
```
samples/net/throughput_server/src/server.c:17:
include/net/buf.h:508:17: error: ‘CONFIG_NET_BUF_USER_DATA_SIZE’ undeclared here (not in a function)
  u8_t user_data[CONFIG_NET_BUF_USER_DATA_SIZE] __net_buf_align;
                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from include/net/net_pkt.h:24:0,
                 from samples/net/throughput_server/src/server.c:17:
samples/net/throughput_server/src/server.c: In function ‘build_reply_pkt’:
samples/net/throughput_server/src/server.c:71:27: error: ‘CONFIG_NET_BUF_DATA_SIZE’ undeclared (first use in this function)
```
Was this some known issue?

tbursztyka commented:
@jukkar for some reason throuhgputserver sample does not enable CONFIG_NET_BUF

tbursztyka commented:
to me this patch set is ready to be merged.

jukkar reviewed (commented):
I will do some quick testing for the PR, will come back shortly.

jukkar commented:
After refactoring this against latest master, my small testing seem to work ok also with gptp, so +1 to that.

jukkar commented on subsys/net/ip/net_pkt.c:
```diff
@@ -396,18 +388,16 @@ struct net_buf *net_pkt_get_reserve_data(struct net_buf_pool *pool,
 
-#if CONFIG_NET_PKT_LOG_LEVEL >= LOG_LEVEL_DBG
-	NET_DBG("%s (%s) [%d] frag %p reserve %u ref %d (%s():%d)",
+#if NET_LOG_LEVEL >= LOG_LEVEL_DBG
```
We need to use CONFIG_NET_PKT_LOG_LEVEL here so that the debug prints are only printed if net_pkt debugging is enabled.  The CONFIG_NET_DEBUG_NET_PKT_ALLOC option depends on/sets NET_LOG_LEVEL so that we get function name/line number info into the alloc/free functions. But we do not want these extra alloc debug prints in that case as they are just printing too much info.

jukkar commented on subsys/net/ip/net_pkt.c:
```diff
@@ -360,22 +354,20 @@ struct net_pkt *net_pkt_get_reserve(struct k_mem_slab *slab,
 	net_pkt_alloc_add(pkt, true, caller, line);
 
 #if NET_LOG_LEVEL == LOG_LEVEL_DBG
```
Noticed that this should have been `CONFIG_NET_PKT_LOG_LEVEL >= LOG_LEVEL_DBG`, can you fix it here

jukkar reviewed (changes requested):
Looks good, minor tweak needed for net_pkt.c

jukkar commented:
No problem, I will send a small fix that is still needed.
